firmware/thruna/BMS/Src/App/App_Eeprom.c

```diff
@@ -114,33 +117,150 @@ EEPROM_StatusTypeDef App_Eeprom_PageErase(struct Eeprom *eeprom, uint16_t page)
     }
 
-    return eeprom->write_page(page, offset, byte_array, num_bytes);
+    return eeprom->write_page(ADDRESS_PAGE, DEFAULT_OFFSET, byte_array, num_bytes);
 }
 
-EEPROM_StatusTypeDef App_Eeprom_Read4CopiesOfAddresses(struct Eeprom *eeprom, uint16_t page, uint16_t *addresses)
+ExitCode App_Eeprom_ReadAddress(struct Eeprom *eeprom, uint16_t *address)
 {
-    uint16_t num_bytes = SAVED_COPIES * sizeof(uint16_t); // saving 3 copies of address, each 2 bytes
-    uint8_t  byte_array[num_bytes];
-    uint8_t  offset = 0;
-    uint8_t  read_status;
+    if (*address >= NUM_PAGES)
+    {
+        return EXIT_CODE_ERROR;
+    }
+
+    uint16_t             num_bytes = SAVED_COPIES * sizeof(uint16_t); // saving 3 copies of address, each 2 bytes
+    uint8_t              byte_array[num_bytes];
+    uint16_t             address_array[SAVED_COPIES];
+    EEPROM_StatusTypeDef read_status;
 
-    read_status = eeprom->read_page(page, offset, byte_array, num_bytes);
+    // Read saved address
+    read_status = eeprom->read_page(ADDRESS_PAGE, DEFAULT_OFFSET, byte_array, num_bytes);
 
+    // Convert read bytes to shorts
     for (uint8_t i = 0; i < SAVED_COPIES; i++)
     {
-        addresses[i] = convert_bytes_to_short(&byte_array[i * sizeof(uint16_t)]);
+        address_array[i] = convert_bytes_to_short(&byte_array[i * sizeof(uint16_t)]);
     }
 
-    return read_status;
+    // Determine if read was successful and if address is uncorrupted
+    if (read_status != EEPROM_OK)
+    {
+        *address = 1;
+        return EXIT_CODE_ERROR;
+    }
+    else if (address_array[0] == address_array[1])
+    {
+        *address = address_array[0];
+        return EXIT_CODE_OK;
+    }
+    else if (address_array[1] == address_array[2])
+    {
+        *address = address_array[1];
+        return EXIT_CODE_OK;
+    }
+    else if (address_array[0] == address_array[2])
+    {
+        *address = address_array[0];
+        return EXIT_CODE_OK;
+    }
+    else if (address_array[0] == address_array[3])
+    {
+        *address = address_array[0];
+        return EXIT_CODE_OK;
+    }
+    else if (address_array[1] == address_array[3])
+    {
+        *address = address_array[1];
+        return EXIT_CODE_OK;
+    }
+    else if (address_array[2] == address_array[3])
+    {
+        *address = address_array[2];
+        return EXIT_CODE_OK;
+    }
+    else
+    {
+        *address = 1;
+        return EXIT_CODE_ERROR;
+    }
+}
+
+EEPROM_StatusTypeDef App_Eeprom_WriteMinSoc(struct Eeprom *eeprom, float min_soc, uint16_t address)
+{
+    float float_arr[SAVED_COPIES];
+
+    for (uint8_t i = 0U; i < SAVED_COPIES; i++)
+    {
+        float_arr[i] = min_soc;
+    }
+
+    return App_Eeprom_WriteFloats(eeprom, address, DEFAULT_OFFSET, float_arr, SAVED_COPIES);
+}
+
+ExitCode App_Eeprom_ReadMinSoc(struct Eeprom *eeprom, uint16_t address, float *min_soc)
+{
+    float                float_arr[SAVED_COPIES];
```

Comment:
Change variable name to something more meaningful. What is float_arr?

Reply:
it's an array of floats to hold the contents read from the EEPROM. EEPROM is written in bytes, so the ReadFloats function reads in bytes and converts them to floats which are written to the input float array. I can update the name